### ISSUE TITLE
When shrinking, try zero value frist as a shortcut

### ISSUE THREAD

**Issue by cmeeren** [hedgehogqa/fsharp-hedgehog#223]
A test may have many generated values. These are potentially large and require many shrinks to get to the "simplest" value (e.g. `0`). Often, a test failure does not depend on many of the generated values. Still, I often experience these not being shrinked fully to the "simplest" value, because Hedgehog gives up shrinking after a while.

This could be alleviated if Hedgehog, upon shrinking, first tested the "simplest" value of one generated parameter. If the test still fails, do this for the next, etc. If the test passes, revert and try the next parameter. After having "shortcutted" all possible parameters like this, which takes only as many shrinks as there are generated parameters, proceed with shrinking as normal. (This would also have the benefit of "saving" most of the shrinking capacity to the generated values that actually need proper shrinking.)

Am I making myself understood, or is my explanation hard to follow?

**Comment by TysonMN** [comment]
> ...Hedgehog gives up shrinking after a while.

When does Hedgehog give up?  Is it after a certain number of shrinks?

**Comment by cmeeren** [comment]
> When does Hedgehog give up? Is it after a certain number of shrinks?

Unsure if that question was for me, but I have no idea.

**Comment by TysonMN** [comment]
It was for whoever knows the answer ;) :P

**Comment by moodmosaic** [comment]
@TysonMN, it'll gave up after 100 discards. @cmeeren, how you'd get the simplest value of a complex data structure?

**Comment by cmeeren** [comment]
> @cmeeren, how you'd get the simplest value of a complex data structure?

Are you talking about e.g. this?

```f#
let myComplexGen =
  gen {
    let! a = Gen.int ...
    let! b = Gen.string ...
    return { A = a; B = b }
  }
```

If so, finding the simplest value for the record is simply to find the simplest value for the generators used when making it (`a` and `b`). In the end, everything boils down to built-in generators, right?

**Comment by TysonMN** [comment]
I don't think it is that simple though because you said
> I often experience these not being shrinked fully to the "simplest" value, because there's a lot to shrink and Hedgehog gives up shrinking after a while.

@moodmosaic said that Hedgehog (only) gives up after 100 discards, so since you are experiencing Hedgehog give up, your generator must include a filter.

I think the behavior you want exists in the generator I created in #224 (since `0` is the second value tested).

**Comment by cmeeren** [comment]
Here's a `DateTimeOffset` generator for dates between 2000 and 2100, centering on the current date (the simplest value). There's no filters here. I have often experienced this not shrinking fully, and a re-run produces a value closer to the base value.

Also, Hedgehog have given up before 100 shrinks (the test report says "after X tests and Y shrinks" where Y < 100; I can't remember ever seeing it report exactly 100 shrinks).

**Comment by TysonMN** [comment]
I will try to reproduce

**Comment by cmeeren** [comment]
Note that in my case, the bug centered on DST changes. Sometimes it would shrink to some DST change in 2040 and thereabouts, whereas other times it would successfully shrink to the DST change closest to the current time.

**Comment by TysonMN** [comment]
> > > ...Hedgehog gives up shrinking after a while.
> > 
> > When does Hedgehog give up? Is it after a certain number of shrinks?
> 
> @TysonMN, it'll gave up after 100 discards. @cmeeren, how you'd get the simplest value of a complex data structure?

Hedgehog is hardcoded to give up after 100 discards...
https://github.com/hedgehogqa/fsharp-hedgehog/blob/81a7502f47cf7704181b87ec4293f1c52c65c96f/src/Hedgehog/Property.fs#L153-L156

...and by default never gives up on shrinking.
https://github.com/hedgehogqa/fsharp-hedgehog/blob/81a7502f47cf7704181b87ec4293f1c52c65c96f/src/Hedgehog/PropertyConfig.fs#L11-L13

@cmeeren, do you recall if your test changed this default shrink limit from `None`?

-----------

However, my guess is that the shrink limit is not the issue.  Especially since you said:
> Also, Hedgehog have given up before 100 shrinks (the test report says "after X tests and Y shrinks" where Y < 100; I can't remember ever seeing it report exactly 100 shrinks).

Hedgehog doesn't always find a smallest failing input.

@cmeeren, do you recall what your test was that used that generated and didn't shrink to the smallest failing input as you expected?

In the tests I added in the recent Hedgehog.Experimential PRs ([1](https://github.com/hedgehogqa/fsharp-hedgehog-experimental/pull/51), [2](https://github.com/hedgehogqa/fsharp-hedgehog-experimental/pull/53), [3](https://github.com/hedgehogqa/fsharp-hedgehog-experimental/pull/54)), I asserted that the shrunken input was _always_ the minimum input that causes the test to fail.  However, this is not possible in general.  I am a bit knowledgeable about how Hedgehog does shrinking, so I know that Hedgehog will always find that minimum failing input (as long at the user code doesn't do something like mutation before the last `let!`/`Gen.bind`).

-----

My guess is that Hedgehog did try the minimal value, but it caused your test to pass.  Then after finding a failing input, it shrunk until it got stuck in a "local optima" instead of finding the minimum failing input.

**Comment by cmeeren** [comment]
> @cmeeren, do you recall if your test changed this default shrink limit from `None`?

It's been a while, but I'm fairly sure I didn't.

> @cmeeren, do you recall what your test was that used that generated and didn't shrink to the smallest failing input as you expected?

No idea, sorry.

**Comment by TysonMN** [comment]
> No idea, sorry.

Then I am thinking that we don't have enough information to investigate further.

Ironically though, your description here is rather consistent with the behavior we are seeing in the draft PR https://github.com/hedgehogqa/fsharp-hedgehog-experimental/pull/55.
